### PR TITLE
perf: compile column regex patterns once per batch, not per row

### PR DESCRIPTION
**Pull request by radmirnovii**
## Summary

- `split`, `regexp_extract`, and `regexp_extract_all` compiled their pattern with `Regex::new` per row whenever the pattern argument is a column (e.g. selected per row by a CASE), and additionally cloned the compiled `Regex` per row — `Regex::clone` does not share the internal lazy-DFA cache, so every row also rebuilt the automaton's transition table.
- A `StrMemo` helper computes any expensive `&str -> T` once per distinct value per batch; the three regex kernels use it and access the compiled regex by reference (`get_or_try_insert_ref`), never by clone.
- Result on a 41M-row column-pattern workload: `regexp_extract` 28.0s → 2.1s, `regexp_extract_all` 24.9s → 3.2s, `split` 5.3s → 3.2s — parity with Spark 4.2, checksums unchanged. Literal patterns also benefit: their scalar arm cloned per row too.

<details>
<summary><b>Mechanism and measurements</b></summary>

## Why per-row compilation happens

When the pattern argument is a column, the kernel receives an array of pattern strings and must produce a `Regex` per row. The pathological but common case is low cardinality: a CASE over a handful of literal patterns materializes a 8192-row batch containing 2–3 distinct strings, and the kernel compiled each row's string independently (~65 µs per `Regex::new` for the benchmark's pattern).

Cloning was the second, subtler cost: profiles of `regexp_extract` over a column pattern showed `Lazy::init_cache` at 19% of CPU, run-time determinization ~17%, determinizer-state hashing ~11%, cache alloc/drop ~5% — the lazy DFA was being rebuilt per row even for an already-compiled regex, because `Regex::clone` starts with an empty DFA cache. Capture extraction itself is not the bottleneck (~2x the cost of `find`).

## `StrMemo`

A per-batch `&str -> T` map keyed by strings borrowed from the input array. It lives for one kernel invocation, so there are no concurrency, growth, or invalidation concerns; it is suitable only for low-cardinality inputs (patterns, formats) — per-row-unique inputs would just pay overhead. `get_or_try_insert_ref` returns `&T` so kernels use the compiled `Regex` in place.

## Measurements

NYC yellow taxi 2024, 12 months (~41M rows), one machine (16 cores / 32 GB), `local[16]`, median of 3 runs after warmup; queries wrapped in a 1-row checksum aggregation, checksums matching Spark 4.2.

| query | before | after | spark 4.2 |
|---|---|---|---|
| regexp_extract, column pattern | 28.0s | 2.1s | 2.0s |
| regexp_extract_all, column pattern | 24.9s | 3.2s | 2.5s |
| split, column pattern | 5.3s | 3.2s | 2.8s |

</details>

<details>
<summary><b>Proposed follow-up</b></summary>

`StrMemo` is an interim measure at the kernel layer. The engine-level replacement is encoding-aware scalar evaluation — dictionary/REE inputs evaluated once per distinct value, plus operators that emit encoded output (e.g. CASE over literals). That mechanism does not exist at the query-engine layer sail builds on today; once it does, `StrMemo` is deleted — every use is one explicit call.

</details>

### CODE REVIEW

**Comment by shehabgamin** [comment]
@radmirnovii Here is an initial review from Claude!

# Review of PR #2448 — perf: compile column regex patterns once per batch

**Why parity holds** (verified against `opt/spark`, not from memory):

- Spark compiles the pattern inside `nullSafeEval` only for fully-non-null rows, with a size-1 last-pattern cache (`RegExpExtractBase.lastRegex`/`getLastMatcher`, `regexpExpressions.scala:831-855`); `StringSplit` compiles per call (`:588-596`). The PR's memo is consulted at the exact statement where the base compiled per row, only in the all-non-null branch — same rows, same order, and an invalid pattern still aborts the batch at the identical row with the identical message.
- Errors are never cached: `get_or_try_insert_ref` runs `compute` before `Entry::Vacant::insert`; Spark's `getPatternAndLastRegex` likewise updates `lastRegex` only on success (`:1247-1258`).
- Column (non-foldable) patterns are genuine Spark behavior for all three expressions (`StringSplit:571`, `RegExpExtract:897`, `RegExpExtractAll:995` — no foldable requirement).
- Byte-keyed memoization matches Spark's `UTF8String`-equality cache key; collation-aware pattern rewriting is identity for the only collation Sail supports.
- The PR body's `Regex::clone` lazy-DFA claim is technically accurate (confirmed against `regex-automata` 0.4.18: `Clone` shares the compiled program, not the `Pool`), and reusing one compiled regex across rows is exactly what Spark's own cache does.
- `codec.rs` unaffected (no UDF added/renamed; all three remain wired at `codec.rs:2849-2851, 3070-3071, 3080`); no concurrency exposure (memo is a per-invocation stack local); every added test assertion pins correct Spark behavior (hand-evaluated).

I compiled and ran a stub replica of the R1 and R2 fixes with standalone `rustc` to validate the borrow/lifetime shapes and behavior — both pass. Line numbers below are PR-head lines.

---

## R1 (minor, optional hardening) — unbounded per-batch memo

**Place on `crates/sail-function/src/functions_utils.rs:59`:**

> `StrMemo` is unbounded: every distinct pattern in the batch keeps its compiled `Regex` alive until the kernel returns. The base code held at most one live compiled `Regex` on the column path (compile → use → drop per row), and Spark itself retains only the last pattern (`RegExpExtractBase.lastRegex`, `regexpExpressions.scala:831-855`), so the worst-case peak-memory profile is new in this PR: an 8192-row batch of distinct adversarial patterns can pin ~8192 compiled programs (each up to ~10 MB at `regex`'s default `size_limit`) per partition. Outputs are unaffected — this is optional hardening, not a correctness issue. Suggest a small capacity cap that falls back to per-row compilation past the cap: it keeps the full win for the CASE-over-a-few-literals case this PR targets while restoring the old worst-case profile.

**Fix:**

```diff
--- a/crates/sail-function/src/functions_utils.rs
+++ b/crates/sail-function/src/functions_utils.rs
@@ -56,12 +56,22 @@ /// formats); per-row-unique inputs just pay overhead.
 pub(crate) struct StrMemo<'a, T> {
     cache: std::collections::HashMap<&'a str, T>,
+    /// Overflow slot used once `cache` is full: past-capacity keys are
+    /// computed per row (the pre-memo behavior) instead of retained, so a
+    /// high-cardinality key column cannot pin one compiled value per
+    /// distinct key for the whole batch.
+    overflow: Option<T>,
 }
 
 impl<'a, T> StrMemo<'a, T> {
+    /// Distinct keys retained per batch: ample for realistic low-cardinality
+    /// pattern columns (a CASE over a handful of literals) while bounding
+    /// worst-case memory; Spark itself retains only the last pattern
+    /// (`RegExpExtractBase.lastRegex`).
+    const CAPACITY: usize = 64;
+
     pub(crate) fn new() -> Self {
         Self {
             cache: std::collections::HashMap::new(),
+            overflow: None,
         }
     }
@@ -78,6 +88,9 @@
     ) -> Result<&T> {
         use std::collections::hash_map::Entry;
+        if self.cache.len() >= Self::CAPACITY && !self.cache.contains_key(key) {
+            return Ok(self.overflow.insert(compute(key)?));
+        }
         match self.cache.entry(key) {
             Entry::Occupied(entry) => Ok(entry.into_mut()),
             Entry::Vacant(entry) => Ok(entry.insert(compute(key)?)),
```

**Test to add** (append inside the existing `mod tests` in the same file):

```diff
@@ -112,4 +125,35 @@
         assert_eq!(calls.get(), 4);
         Ok(())
     }
+
+    #[test]
+    fn str_memo_recomputes_past_capacity_keys_and_still_hits_retained_keys() -> Result<()> {
+        let capacity = StrMemo::<'static, usize>::CAPACITY;
+        let keys: Vec<String> = (0..=capacity).map(|i| "k".repeat(i + 1)).collect();
+        let calls = Cell::new(0);
+        let mut memo: StrMemo<'_, usize> = StrMemo::new();
+        let compute = |s: &str| {
+            calls.set(calls.get() + 1);
+            Ok(s.len())
+        };
+        for key in keys.iter().take(capacity) {
+            assert_eq!(*memo.get_or_try_insert_ref(key, compute)?, key.len());
+        }
+        assert_eq!(calls.get(), capacity);
+        let past_capacity = &keys[capacity];
+        assert_eq!(
+            *memo.get_or_try_insert_ref(past_capacity, compute)?,
+            past_capacity.len()
+        );
+        assert_eq!(
+            *memo.get_or_try_insert_ref(past_capacity, compute)?,
+            past_capacity.len()
+        );
+        // A past-capacity key is recomputed on each call, not retained.
+        assert_eq!(calls.get(), capacity + 2);
+        // Keys retained before the cache filled still hit without recomputing.
+        assert_eq!(*memo.get_or_try_insert_ref(&keys[0], compute)?, 1);
+        assert_eq!(calls.get(), capacity + 2);
+        Ok(())
+    }
 }
```

---

## R2 (minor) — scalar-vs-memo resolution block duplicated ×3 (rationale comment ×4)

**Place on `crates/sail-function/src/scalar/string/spark_regexp_extract_all.rs:168`:**

> This scalar-vs-memo match (and the two comment lines above it) now appears three times — here, at L237, and in `spark_split.rs` L146 — and the "By reference / lazy-DFA" rationale is a fourth near-copy of the doc on `get_or_try_insert_ref`. Consider hoisting a `StrMemo::resolve(scalar, key, compute)` helper so the rationale lives once on the helper's doc. One constraint: the key must be a closure evaluated only on the memoized path — the pattern argument is `Hint::AcceptsSingular`, so a scalar pattern stays a length-1 array while `i` ranges over the batch, and an eager `pattern.value_(i)` argument would read out of bounds.

**Fix — helper** (`functions_utils.rs`, after `get_or_try_insert_ref`):

```diff
@@ -82,6 +82,25 @@
             Entry::Vacant(entry) => Ok(entry.insert(compute(key)?)),
         }
     }
+
+    /// Resolves the value for one row: a pre-compiled scalar if there is
+    /// one, otherwise the per-batch memoized value for the row's key. `key`
+    /// is evaluated only on the memoized path, so scalar inputs may be
+    /// length-1 arrays while the row index ranges over the whole batch. The
+    /// value is returned by reference: cloning a `Regex` per row would
+    /// discard its internal lazy-DFA cache and rebuild the automaton on
+    /// every match.
+    pub(crate) fn resolve<'m>(
+        &'m mut self,
+        scalar: Option<&'m T>,
+        key: impl FnOnce() -> &'a str,
+        compute: impl FnOnce(&str) -> Result<T>,
+    ) -> Result<&'m T> {
+        match scalar {
+            Some(value) => Ok(value),
+            None => self.get_or_try_insert_ref(key(), compute),
+        }
+    }
 }
```

**Fix — the three call sites:**

```diff
--- a/crates/sail-function/src/scalar/string/spark_regexp_extract_all.rs
+++ b/crates/sail-function/src/scalar/string/spark_regexp_extract_all.rs
@@ -163,15 +163,12 @@
                 if pattern_is_null || idx_is_null || values.is_null(i) {
                     builder.append_null();
                 } else {
-                    // By reference: cloning a Regex per row would discard its
-                    // lazy-DFA cache and rebuild the automaton on every match.
-                    let re = match pattern_scalar_opt.as_ref() {
-                        Some(re) => re,
-                        None => regex_memo.get_or_try_insert_ref(pattern.value_(i), |p| {
-                            parse_regex(SparkRegexpExtract::NAME, p)
-                        })?,
-                    };
+                    let re = regex_memo.resolve(
+                        pattern_scalar_opt.as_ref(),
+                        || pattern.value_(i),
+                        |p| parse_regex(SparkRegexpExtract::NAME, p),
+                    )?;
                     let group_idx = idx_scalar_opt.unwrap_or_else(|| idx.value(i));
@@ -232,15 +229,12 @@
                 if pattern_is_null || idx_is_null || values.is_null(i) {
                     builder.append_null();
                 } else {
-                    // By reference: cloning a Regex per row would discard its
-                    // lazy-DFA cache and rebuild the automaton on every match.
-                    let re = match pattern_scalar_opt.as_ref() {
-                        Some(re) => re,
-                        None => regex_memo.get_or_try_insert_ref(pattern.value_(i), |p| {
-                            parse_regex(SparkRegexpExtractAll::NAME, p)
-                        })?,
-                    };
+                    let re = regex_memo.resolve(
+                        pattern_scalar_opt.as_ref(),
+                        || pattern.value_(i),
+                        |p| parse_regex(SparkRegexpExtractAll::NAME, p),
+                    )?;
                     let group_idx = idx_scalar_opt.unwrap_or_else(|| idx.value(i));
--- a/crates/sail-function/src/scalar/string/spark_split.rs
+++ b/crates/sail-function/src/scalar/string/spark_split.rs
@@ -141,14 +141,11 @@
                 {
                     builder.append_null();
                 } else {
-                    // By reference: cloning a Regex per row would discard its
-                    // lazy-DFA cache and rebuild the automaton on every match.
-                    let format_regex = match format_scalar_opt.as_ref() {
-                        Some(format_regex) => format_regex,
-                        None => regex_memo.get_or_try_insert_ref(format.value(i), parse_regex)?,
-                    };
+                    let format_regex = regex_memo.resolve(
+                        format_scalar_opt.as_ref(),
+                        || format.value(i),
+                        parse_regex,
+                    )?;
                     let limit = limit_scalar_opt.unwrap_or_else(|| limit.value(i));
```

The removed comments are new in this PR and their content moves verbatim into `resolve`'s doc — no pre-existing comment is lost; the one-line "Compile each distinct column pattern once per batch" comments at the three `StrMemo::new()` sites stay. If the maintainers judge the two-closure signature worse than the explicit match, the minimum alternative is comment-only: keep the rationale solely on `get_or_try_insert_ref`'s doc and shrink each call-site comment to a one-line pointer.

---

## R3 (minor) — no end-to-end BDD coverage of column patterns for the two regexp functions

**Place on `crates/sail-function/src/scalar/string/spark_regexp_extract_all.rs:412`:**

> These kernel tests are consistent with crate precedent, but the project default for SQL-observable behavior is BDD (`dev.instructions.md:75`), and neither `regexp_extract.feature` nor `regexp_extract_all.feature` exercises a pattern supplied by a column anywhere — the exact path this PR changes. (`split.feature` L78 already covers it for `split`, including repeated patterns.) Suggest adding the two scenarios below; the repeated-pattern rows also pin the memo-hit path end-to-end, and the NULL-pattern row pins null-intolerant propagation.

**Test — `python/pysail/tests/spark/function/features/string/regexp_extract.feature`** (after the "No match and edge cases" examples at line 55, before `@function(nullability)`):

```diff
@@ -53,6 +53,21 @@
         | regexp_extract with anchored pattern at end       | 'abc123' | '(\\d+)$' | 123    |
 
+  Rule: Pattern from a column
+
+    Scenario: regexp_extract with the pattern supplied by a column
+      When query
+        """
+        SELECT regexp_extract(s, p, 1) AS result FROM VALUES ('a11b', '([0-9]+)'), ('a22b', '([0-9]+)'), ('a33b', '([a-z]+)'), ('a44b', CAST(NULL AS STRING)) AS t(s, p)
+        """
+      Then query result
+        | result |
+        | 11     |
+        | 22     |
+        | a      |
+        | NULL   |
+
   @function(nullability)
   Rule: Output schema
```

**Test — `python/pysail/tests/spark/function/features/string/regexp_extract_all.feature`** (after the "NULL handling" examples at line 74):

```diff
@@ -72,6 +72,21 @@
         | regexp_extract_all returns NULL when pattern is NULL | 'abc', NULL    |
 
+  Rule: Pattern from a column
+
+    Scenario: regexp_extract_all with the pattern supplied by a column
+      When query
+        """
+        SELECT regexp_extract_all(s, p, 1) AS result FROM VALUES ('1a2b', '([0-9])'), ('3c4d', '([0-9])'), ('3c4d', '([a-z])'), ('5e6f', CAST(NULL AS STRING)) AS t(s, p)
+        """
+      Then query result
+        | result |
+        | [1, 2] |
+        | [3, 4] |
+        | [c, d] |
+        | NULL   |
+
   @function(nullability)
   Rule: Output schema
```

Expected values hand-checked against Spark's `RegExpExtract`/`RegExpExtractAll` `nullSafeEval`; rows are distinct so the steps' default unordered comparison is safe, and list/NULL rendering matches the files' existing conventions.

---

## R4 (nit) — memo-hit path untested in two of the three kernel tests

**Place on `crates/sail-function/src/scalar/string/spark_split.rs:195`:**

> The repeated `"-"` here looks like it exercises the memo-hit path, but row 2's *value* is null, so that row short-circuits to `append_null()` before the memo lookup — this test only ever takes memo misses. Suggest a fourth row that reuses `"-"` with a non-null value; while here, `first` never asserts `value(1)`.

**Place on `crates/sail-function/src/scalar/string/spark_regexp_extract_all.rs:453`:**

> Both patterns are distinct, so this test only exercises memo misses; of the three kernels only `extract_with_column_pattern` actually hits the memo. Suggest a third row reusing `"[0-9]"` so the hit path — the point of this PR — is pinned in this kernel too.

**Fix — `spark_split.rs` test:**

```diff
@@ -190,13 +190,24 @@
     /// A delimiter supplied as a column selects the split pattern per row;
-    /// null inputs yield null outputs.
+    /// null inputs yield null outputs; a repeated delimiter takes the
+    /// memoized path.
     #[test]
     fn split_with_column_pattern() -> Result<()> {
-        let values: ArrayRef = Arc::new(StringArray::from(vec![Some("a-b-c"), Some("a:b"), None]));
-        let formats: ArrayRef = Arc::new(StringArray::from(vec![Some("-"), Some(":"), Some("-")]));
-        let limits: ArrayRef = Arc::new(Int32Array::from(vec![-1, -1, -1]));
+        let values: ArrayRef = Arc::new(StringArray::from(vec![
+            Some("a-b-c"),
+            Some("a:b"),
+            None,
+            Some("x-y"),
+        ]));
+        let formats: ArrayRef = Arc::new(StringArray::from(vec![
+            Some("-"),
+            Some(":"),
+            Some("-"),
+            Some("-"),
+        ]));
+        let limits: ArrayRef = Arc::new(Int32Array::from(vec![-1, -1, -1, -1]));
         let output = spark_split_inner(&[values, formats, limits])?;
         let output = output
             .as_any()
             .downcast_ref::<ListArray>()
             .expect("list output");
-        assert_eq!(output.len(), 3);
+        assert_eq!(output.len(), 4);
         let first = output.value(0);
         let first = first
             .as_any()
             .downcast_ref::<StringArray>()
             .expect("string parts");
-        assert_eq!((first.len(), first.value(0), first.value(2)), (3, "a", "c"));
+        assert_eq!(
+            (first.len(), first.value(0), first.value(1), first.value(2)),
+            (3, "a", "b", "c")
+        );
         let second = output.value(1);
@@ -216,6 +227,15 @@
         assert!(output.is_null(2));
+        // Row 3 reuses "-" after row 0 compiled it: the memoized regex must
+        // split exactly as a fresh compile would.
+        let fourth = output.value(3);
+        let fourth = fourth
+            .as_any()
+            .downcast_ref::<StringArray>()
+            .expect("string parts");
+        assert_eq!((fourth.len(), fourth.value(0), fourth.value(1)), (2, "x", "y"));
         Ok(())
     }
 }
```

**Fix — `spark_regexp_extract_all.rs` test:**

```diff
@@ -450,11 +450,11 @@
     #[test]
     fn extract_all_with_column_pattern() -> Result<()> {
-        let values = strings(vec![Some("1a2"), Some("3b4")]);
-        let patterns = strings(vec![Some("[0-9]"), Some("[a-z]")]);
-        let idx: ArrayRef = Arc::new(Int64Array::from(vec![0i64, 0]));
+        let values = strings(vec![Some("1a2"), Some("3b4"), Some("5c6")]);
+        let patterns = strings(vec![Some("[0-9]"), Some("[a-z]"), Some("[0-9]")]);
+        let idx: ArrayRef = Arc::new(Int64Array::from(vec![0i64, 0, 0]));
         let output = regexp_extract_all_inner(&[values, patterns, idx])?;
@@ -473,6 +473,17 @@
         assert_eq!((letters.len(), letters.value(0)), (1, "b"));
+        // Row 2 reuses "[0-9]" after row 0 compiled it: the memoized regex
+        // must extract exactly as a fresh compile would.
+        let reused = output.value(2);
+        let reused = reused
+            .as_any()
+            .downcast_ref::<StringArray>()
+            .expect("string matches");
+        assert_eq!(
+            (reused.len(), reused.value(0), reused.value(1)),
+            (2, "5", "6")
+        );
         Ok(())
     }
 }
```

---

## R5 (nit) — inline `std::collections` paths against crate convention

**Place on `crates/sail-function/src/functions_utils.rs:79`:**

> Nit: the crate convention is file-top imports (`spark_date_format.rs:1-2` imports both `HashMap` and `hash_map::Entry` at the top); the inline `std::collections::` paths here and at L59/L65 are the only ones in `sail-function`.

**Fix:**

```diff
--- a/crates/sail-function/src/functions_utils.rs
+++ b/crates/sail-function/src/functions_utils.rs
@@ -1,4 +1,6 @@
 /// [Credit]: <https://github.com/apache/datafusion/blob/e6e1eb229440591263c82bb2b913a4d5a16f9b70/datafusion/functions/src/utils.rs>
+use std::collections::HashMap;
+use std::collections::hash_map::Entry;
 use std::sync::Arc;
@@ -57,7 +59,7 @@
 pub(crate) struct StrMemo<'a, T> {
-    cache: std::collections::HashMap<&'a str, T>,
+    cache: HashMap<&'a str, T>,
 }
@@ -63,7 +65,7 @@
     pub(crate) fn new() -> Self {
         Self {
-            cache: std::collections::HashMap::new(),
+            cache: HashMap::new(),
         }
     }
@@ -78,7 +80,6 @@
     ) -> Result<&T> {
-        use std::collections::hash_map::Entry;
         match self.cache.entry(key) {
```

Note: the diffs for R1, R2, and R5 all touch `functions_utils.rs` and are each written standalone against the PR head, so if you apply more than one, apply them in that order (or hand-merge — R1 modifies the struct and method body, R2 appends a method, R5 rewrites the paths R1's context lines mention).

---

File: crates/sail-function/src/functions_utils.rs [hunk 1]
```diff
@@ -50,3 +50,66 @@ where
         }
     })
 }
+
+/// Per-batch memoization of an expensive `&str -> T` computation (a compiled
+/// regex, a parsed interval). Keys borrow from the input, so a cache lives one
+/// kernel invocation. Only worth it for low-cardinality inputs (patterns,
+/// formats); per-row-unique inputs just pay overhead.
+pub(crate) struct StrMemo<'a, T> {
```
[the remark this answers is not in the document]
Comment (radmirnovii) [reply crates/sail-function/src/functions_utils.rs:68]: Confirmed — a worst-case batch pins ~612 MiB of compiled `Regex` state. Added a hard cap (`MAX_DISTINCT = 128`): past it, values are computed per call and not cached.

**Comment by radmirnovii** [comment]
Thanks for the thorough review. R1 was already addressed in 23eecc98 — same overflow-slot shape you suggested, with the cap set to `MAX_DISTINCT = 128` sized from a measured ~75 KiB per used `Regex`. R2–R5 applied in 6cbd42a2 and 925deeb6: `StrMemo::resolve` with the lazy key closure, file-top imports, memo-hit rows in the split/extract_all kernel tests, and the two column-pattern BDD scenarios — expected values verified against JVM Spark 4.2 via `SPARK_REMOTE`, not only hand-checked.

**Review by shehabgamin — APPROVED** [review]
Looks great! Thanks so much!
